### PR TITLE
First class support for unbounded window function bounds

### PR DESCRIPTION
**Pull request by mythrocks**
This change is in relation to https://github.com/NVIDIA/spark-rapids/issues/1039.

CUDF  does not currently support the notion of an `UNBOUNDED` window frame, i.e. a way to specify `UNBOUNDED PRECEDING` or `UNBOUNDED FOLLOWING` boundary. This is faked by the caller by specifying `std::numeric_limits<size_type>::max` for either bounds.

This workaround is acceptable for `ROW`-based window frames, but not for `RANGE`-based queries such as those supported by `cudf::grouped_time_range_rolling_window()`.

For instance, consider the following dataset:
```
// For this illusion, assume the following:
//   1. timestamp column values to be, say, the day in January 2020.
//   2. ∅ == null

Group-by column    (gby):   [0, 0, 0, 0, 0,  1, 1, 1, 1, 1]
Aggregation column (agg):   [0, 1, 2, 3, 4,  5, 6, 7, 8, 9]
Timestamp column    (ts):   [∅, ∅, ∅, 1, 2,  ∅, ∅, 3, 4, 5]
```
As per the changes in #6557, a `COUNT(agg)` over time-range rolling_window `(5000 DAYS PRECEDING, 1 DAY FOLLOWING)` would yield:
```
Count over (5000d, 1d):     [3, 3, 3, 2, 2,  2, 2, 2, 3, 3]
```
This is understandable, since `NULL` timestamps cannot participate in timestamp subtraction. However, for a truly `UNBOUNDED PRECEDING` window, the expected result (congruent with the semantics of SQL and most big data analytics systems) would be:
```
Count over (UNBOUNDED, 1d): [3, 3, 3, 5, 5,  2, 2, 4, 5, 5]
```
For unbounded windows, there are no window-boundary comparisons required. The window may freely extend into the null-timestamp group.

The commits in this PR aim to add explicit support for `UNBOUNDED` window boundaries, for use with "grouped" rolling window functions.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by harrism** [comment]
Please use draft PRs while WIP so it's obvious from the github notifications view (which doesn't show labels).

**Comment by mythrocks** [comment]
> Please use draft PRs while WIP so it's obvious from the github notifications view (which doesn't show labels).

Oh, right. Sorry.

**Review by harrism — APPROVED** [review]
I'm convinced that "rolling windows" is what happens  when people generalize computing needs into APIs rather than programming languages. What a complete mess.

But hey, it wasn't your idea.  This PR looks clean. I approve.

**Comment by mythrocks** [comment]
> What a mess (I mean the conglomeration of features, not the code.)

"This isn't even my final form."

File: cpp/src/rolling/rolling.cu [hunk 1]
```diff
@@ -1525,6 +1584,8 @@ std::unique_ptr<column> time_range_window_ASC(
     auto nulls_begin = d_nulls_begin[group_label];
     auto nulls_end   = d_nulls_end[group_label];
 
+    if (following_window_is_unbounded) { return group_end - idx - 1; }
```
Comment (nvdbaranec) [inline cpp/src/rolling/rolling.cu:1587]: Nitpick:  should always err on the side of parentheses with compound arithmetic.

File: cpp/src/rolling/rolling.cu [hunk 2]
```diff
@@ -1604,11 +1674,14 @@ std::unique_ptr<column> time_range_window_DESC(column_view const& input,
...
-                               d_timestamps = timestamp_column.data<TimeT>(),
-                               following_window] __device__(size_type idx) -> size_type {
+  auto following_calculator =
+    [nulls_begin_idx,
+     nulls_end_idx,
+     num_rows     = input.size(),
+     d_timestamps = timestamp_column.data<TimeT>(),
+     following_window,
+     following_window_is_unbounded] __device__(size_type idx) -> size_type {
+    if (following_window_is_unbounded) { return num_rows - idx - 1; }
```
Comment (nvdbaranec) [inline cpp/src/rolling/rolling.cu:1684]: Same comment on parentheses as above.

File: cpp/src/rolling/rolling.cu [hunk 3]
```diff
@@ -1662,17 +1737,21 @@ std::unique_ptr<column> time_range_window_DESC(
...
+     d_nulls_begin   = null_start.data().get(),
+     d_nulls_end     = null_end.data().get(),
+     preceding_window,
+     preceding_window_is_unbounded] __device__(size_type idx) -> size_type {
     auto group_label = d_group_labels[idx];
     auto group_start = d_group_offsets[group_label];
     auto nulls_begin = d_nulls_begin[group_label];
     auto nulls_end   = d_nulls_end[group_label];
 
+    if (preceding_window_is_unbounded) { return idx - group_start + 1; }
```
Comment (nvdbaranec) [inline cpp/src/rolling/rolling.cu:1753]: Parentheses

File: cpp/src/rolling/rolling.cu [hunk 4]
```diff
@@ -1698,18 +1777,22 @@ std::unique_ptr<column> time_range_window_DESC(
...
+     d_nulls_end     = null_end.data().get(),
+     following_window,
+     following_window_is_unbounded] __device__(size_type idx) -> size_type {
     auto group_label = d_group_labels[idx];
     auto group_start = d_group_offsets[group_label];
     auto group_end   = d_group_offsets[group_label + 1];
     auto nulls_begin = d_nulls_begin[group_label];
     auto nulls_end   = d_nulls_end[group_label];
 
+    if (following_window_is_unbounded) { return group_end - idx - 1; }
```
Comment (nvdbaranec) [inline cpp/src/rolling/rolling.cu:1794]: Parentheses.

**Review by nvdbaranec — CHANGES_REQUESTED** [review]
Couple of tiny nitpicks.  And one additional question :  I'm assuming we're leaving the old size_type based bounding interfaces in there for backwards compatibility?

File: cpp/include/cudf/rolling.hpp [hunk 5]
```diff
@@ -83,6 +83,42 @@ std::unique_ptr<column> rolling_window(
...
+  static window_bounds get(size_type value) { return window_bounds(false, value); }
+
+  /**
+   * @brief Construct unbounded window boundary.
+   *
+   * @return window_bounds
+   */
+  static window_bounds unbounded()
+  {
+    return window_bounds(true, std::numeric_limits<cudf::size_type>::max());
```
[the remark this answers is not in the document]
Comment (mythrocks) [reply cpp/include/cudf/rolling.hpp:105]: > Or is it the case that for range based bounds we check the `is_unbounded` bool?

Yes, that's right. It should be correct to use the bounds value for non-range queries.

**Comment by mythrocks** [comment]
> I'm assuming we're leaving the old size_type based bounding interfaces in there for backwards compatibility?

Yes, that is right; these are public APIs.

Also, I am leaving the `rolling_window()` functions unchanged for the moment. These overloads can be introduced at a later time, if the Pandas side requires them.

**Review by revans2 — APPROVED** [review]

**Comment by mythrocks** [comment]
Found a bug in the `WindowOptions.hashCode()`. (I neglected to account for the new unbounded flags.)

I have added tests for all combinations of `UNBOUNDED`, `PRECEDING`/`FOLLOWING`/`CURRENT ROW` on the JNI side, in case I've missed something else. Could I request @revans2 to please take a look at the new changes?

**Review by revans2 — APPROVED** [review]

**Comment by mythrocks** [comment]
This has been merged. Thank you for the reviews, all.

**Pull request merged**